pkg/util/validation/formats.go

```diff
@@ -209,6 +216,135 @@ func (FormatHTTPOriginSpec) CheckFormat(value interface{}) error {
 	return nil
 }
 
+type FormatLDAPURL struct{}
+
+func (FormatLDAPURL) CheckFormat(value interface{}) error {
+	str, ok := value.(string)
+	if !ok {
+		return nil
+	}
+
+	u, err := url.Parse(str)
+	if err != nil {
+		return err
+	}
+
+	if u.Scheme != "ldap" && u.Scheme != "ldaps" {
+		return errors.New("expect input URL with scheme ldap / ldaps")
+	}
+
+	if u.Host == "" {
+		return errors.New("expect input URL with non-empty host")
+	}
+
+	err = errors.New("expect input URL without user info, path, query and fragment")
+	if u.User != nil {
+		return err
+	}
+	if u.Path != "" || u.RawPath != "" {
+		return err
+	}
+	if u.RawQuery != "" {
+		return err
+	}
+	if u.Fragment != "" || u.RawFragment != "" {
+		return err
+	}
+
+	return nil
+
+}
+
+type FormatLDAPDN struct{}
+
+func (FormatLDAPDN) CheckFormat(value interface{}) error {
+	str, ok := value.(string)
+	if !ok {
+		return nil
+	}
+
+	if str == "" {
+		return errors.New("expect non-empty base DN")
+	}
+
+	dn, err := ldap.ParseDN(str)
+	if err != nil {
+		return err
+	}
+	if len(dn.RDNs) == 0 {
+		return errors.New("invalid DN")
+	}
+
+	return nil
+
+}
+
+type FormatLDAPSearchFilterTemplate struct{}
+
+func (FormatLDAPSearchFilterTemplate) CheckFormat(value interface{}) error {
+	str, ok := value.(string)
+	if !ok {
+		return nil
+	}
+
+	tmpl, err := template.New("search_filter").Parse(str)
+	if err != nil {
+		return err
+	}
+	var buf bytes.Buffer
+	// make a string array of phone, username, email
+	testcases := []string{"+85298765432", "username", "test@test.com"}
+	for _, testcase := range testcases {
+		err = tmpl.Execute(&buf, map[string]string{"Username": testcase})
+		if err == nil && buf.String() != "" {
+			break
```

Comment:
The search filter template MUST pass all three test cases in order to be considered as valid. So the loop should look like

```golang
testcases := []string{"+85298765432", "username", "test@test.com"}
for _, testcase := range testcases {
  var buf bytes.Buffer // Note that buf is moved into the loop. You do not need it outside the loop.
  err := tmpl.Execute(&buf, map[string]string{"Username": testcase})
  if err != nil {
    return err
  }
  filterString := buf.String()
  _, err := ldap.CompileFilter(filterString)
  if err != nil {
    return err
  }
}

// If no early return, then the filter template is considered as valid.
return nil
```